Add documentation for spellchecker and spellcheck docs

zslade opened the pull request:
### Type of PR

- [ ] BUG
- [ ] FEAT
- [x] MAINT
- [x] DOC

### Is your Pull Request linked to an existing Issue or Pull Request?
<!--
  Add links to related issues/prs. For Example "closes #111"
-->
Docs spellchecker update: #2000 
Original spellchecker PR: #1588 

### Give a brief description for the solution you have provided
<!--
  Provide a clear and concise description of what you want to happen.
-->

I have updated the documentation with guidance on how to use the docs spellchecker, following on from this issue #2000.

All docs have been spellchecked to set a baseline for 'spellchecked' docs. This involves
1. adding legitimate spellings to `custom_dictonary.txt` which aren't being picked up by LibreOffice dictionaries, e.g. words like Splink
2. adding regex to `pyspelling.yml` to bypass certain patterns or sections in the docs, e.g. code blocks

@ThomasHepworth and I decided not to make the spellchecker part of CD/CI with a GitHub Action at this stage as would require some non-trivial configuration and a new check box on the PR template will probably suffice.

<details>
  <summary>Code for testing ignore rules in pyspelling.yml</summary>

# General text

This is some text that does not contain a mistake

<!-- spelling mistake to check spellchecker hitting start of file -->
startoffilemistake

# colons

<!-- colon with no space after, words, then colon -->
With a massive thanks to external contributor [@hanslemm](https://github.com/moj-analytical-services/splink/pull/1191), Splink now supports :simple-postgresql: Postgres. To get started, check out the [Postgres Topic Guide](https://moj-analytical-services.github.io/splink/topic_guides/backends/postgres.html).

<!-- above rule will catch these too -->
[:colon-mistake: RSS feed]
:octicons-duplicate-24-mistake:


## Code block ignore test

<!-- Doesn't seem to have a problem with code blocks or inline code -->

```

Code block mssstake

```

## python code

```py

  pycodeblockmistakeee

```

## alt python blocks

```python
import jellyfish
altpythonblockmistake
jellyfish.jaro_similarity("MARTHA", "AMRTHA")

```

## inline code

`inlinecodemistkakeeee`

`l.first_name = r.first_name and substr(l.surname,1,1) = substr(r.surname,1,1)`.

## blog post headers
---
date: 2024-01-23
authors:
  - zoe-s
  - alice-o
categories:
  - Ethics
---

## anchor tags

<a href="https://t.co/eXSNHHe2kc">anchormistake</a>

## Test text between `diff` blocks

??? info "Example of convergence output"

    ```diff
    @@ -1,9 +1,8 @@
    {
    diffblockmistake
    -  
    ```

## links

[`create_function` function](https://docs.python.org/3/library/sqlite3.html#sqlite3.Connection.create_function)

<!-- testspellone -->

## text between `$$`

$$\text{}
mathcodeblockmistake
\something $$

<!-- testspelltwo -->

Almost 100%, say 98% $inlinemathmistake m \approx 0.98$

## compound words

Equi-join

## Card content ignore test

::cards::
[
    {
      cardblockmistakee
    },
]
::/cards::

## auto generated code

::: splink.linker.Linker
    handler: python
    selection:
      members:
        - cluster_pairwise_predictions_at_threshold
        - compare_two_records
        - compute_tf_table
        - deterministic_link
        - find_matches_to_new_records
        - load_settings
        - load_model
        - load_settings_from_json
        - predict
    rendering:
      show_root_heading: false
      show_source: true

## BibTeX blocks

```BibTeX
@article{Linacre_Lindsay_Manassis_Slade_Hepworth_2022,
	title        = {Splink: Free software for probabilistic record linkage at scale.},
	author       = {Linacre, Robin and Lindsay, Sam and Manassis, Theodore and Slade, Zoe and Hepworth, Tom and Kennedy, Ross and Bond, Andrew},
	year         = 2022,
	month        = {Aug.},
	journal      = {International Journal of Population Data Science},
	volume       = 7,
	number       = 3,
	doi          = {10.23889/ijpds.v7i3.1794},
	url          = {https://ijpds.org/article/view/1794},
}
```

## error logger

To enable the logging of multiple errors in a singular check, or across multiple checks, an [`ErrorLogger`](https://github.com/moj-analytical-services/splink/blob/settings_validation_refactor_and_improved_logging/splink/exceptions.py#L34) class is available for use.

The `ErrorLogger` operates in a similar way to working with a list, allowing you to add additional errors using the `append` method. Once you've logged all of your errors, you can raise them with the `raise_and_log_all_errors` method.

??? note "`ErrorLogger` in practice"
    ```py
    from splink.exceptions import ErrorLogger

    # Create an error logger instance
    e = ErrorLogger()

    # Log your errors
    e.append(SyntaxError("The syntax is wrong"))
    e.append(NameError("Invalid name entered"))

    # Raise your errors
    e.raise_and_log_all_errors()
    ```

<!-- Spelling error to check reaching end of file -->
endoffilemistake  

</details>


### PR Checklist

- [x] Added documentation for changes
- [ ] Added feature to example notebooks or tutorial (if appropriate)
- [ ] Added tests (if appropriate)
- [ ] Updated CHANGELOG.md (if appropriate)
- [x] Made changes based off the latest version of Splink
- [x] Run the [linter](https://moj-analytical-services.github.io/splink/dev_guides/changing_splink/lint.html)

zslade commented:
> Note: I briefly messed around with the spellchecker, and wasn't immediately able to get it working on Linux. It looks like the script as it currently stands is Mac-only, but even manually installing Aspell, I couldn't get it to use `.aff` and `.dic` files.

Thanks @zmbc for drawing this to our attention! Will have a look into it :)

zslade commented:
> Note: I briefly messed around with the spellchecker, and wasn't immediately able to get it working on Linux. It looks like the script as it currently stands is Mac-only, but even manually installing Aspell, I couldn't get it to use `.aff` and `.dic` files.

Thanks again @zmbc. Can I just check which Linux distribution and version you're running? I'm assuming you tried to run the spellchecker following the instructions in this PR - is that correct? If so, it is probably because extra dependencies are required for Linux, e.g. we've been able to run the spellchecker on a Docker file using alpine 3.14 but this required installing some additional dependencies: `aspell-libs` and `aspell-en` (as well as `aspell`)

zslade commented:
> @zslade Sorry -- I should have given a bit more information. I'm on Ubuntu 20.04 via Windows Subsystem for Linux. I installed aspell via conda. I _can_ run the spellchecker script (if I comment out the Mac-specific part) but I have no reason to believe that it is using the LibreOffice dictionaries. `./scripts/pyspelling/spellchecker.sh | wc -l` outputs `797` -- so it is finding a lot of "typos" in the docs as they currently are on the `splink4_dev` branch.

Thanks @zmbc. Just so I'm really clear - the spellchecker runs for you (spellchecks the docs) but doesn't pass (finds spelling errors)? We are still in the process of updating `custom_dictonary.txt` with legitimate spellings which might not be found in the LibreOffice dictionaries e.g. Splink so it might be the case that everything is working as intended but there is still a fair amount of spelling correcting/dictionary updating to do! 🙏  This is being done as part of this PR - I will update the description to reflect that.

As a rough check, could you please `git checkout spellchecker_docs` and run `./scripts/pyspelling/spellchecker.sh | wc -l` again? The resulting line count should be significantly lower than that for `splink4_dev` or `master` as exception patterns have been added to `pyspelling.yml` and words have been added to `custom_dictonary.txt`.

Would you also be able to tell us which Mac-specific parts you commented out? This will help us if we decide built a system-agnostic version :)

Many thanks!

zmbc commented:
> As a rough check, could you please git checkout spellchecker_docs and run ./scripts/pyspelling/spellchecker.sh | wc -l again?

I get `282`. The full output:

<details>

```
Path to spellcheck: docs
============== Running pyspelling spellchecker on docs ==============
docs/blog/index.md docs/charts/index.md docs/dev_guides/caching.md docs/dev_guides/debug_modes.md docs/dev_guides/dependency_management.md docs/dev_guides/index.md docs/dev_guides/spark_pipelining_and_caching.md docs/dev_guides/transpilation.md docs/dev_guides/udfs.md docs/includes/tags.md docs/settingseditor/editor.md docs/topic_guides/topic_guides_index.md
Misspelled words:
<context> docs/blocking_rule_composition.md
--------------------------------------------------------------------------------
splink
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/includes/generated_files/comparison_library_dialect_table.md
--------------------------------------------------------------------------------
PostgreSql
damerau
jaccard
jaro
levenshtein
winkler
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/includes/generated_files/comparison_composition_library_dialect_table.md
--------------------------------------------------------------------------------
PostgreSql
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/includes/generated_files/comparison_level_library_dialect_table.md
--------------------------------------------------------------------------------
PostgreSql
damerau
jaccard
jaro
levenshtein
winkler
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/includes/generated_files/comparison_template_library_dialect_table.md
--------------------------------------------------------------------------------
PostgreSql
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/includes/generated_files/datasets_table.md
--------------------------------------------------------------------------------
splink
wikidata
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/linkerpred.md
--------------------------------------------------------------------------------
tf
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/getting_started.md
--------------------------------------------------------------------------------
PostgreSql
conda
duckdbless
github
splink
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/dev_guides/changing_splink/contributing_to_docs.md
--------------------------------------------------------------------------------
pyspelling
yml
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/spelling_test_file.md
--------------------------------------------------------------------------------
Longrightarrow
endoffilemistake
inlinecodemista
pymistake
startoffilemistake
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/documentation_index.md
--------------------------------------------------------------------------------
autocomplete
autoformatting
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/topic_guides/data_preparation/feature_engineering.md
--------------------------------------------------------------------------------
ATSN
Bedfordshire
ComparisonLevels
DF
DuckDBLinker
FTSN
HU
Harpenden
JL
NG
QE
QF
TLR
TQ
acos
chudleigh
clifford
devon
dm
dmetaphone
geonames
geospatial
jaro
nan
oNah
splink
subdistricts
substrings
thomas
winkler
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/topic_guides/theory/fellegi_sunter.md
--------------------------------------------------------------------------------
Longrightarrow
frac
textsf
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/topic_guides/evaluation/model.md
--------------------------------------------------------------------------------
frac
textsf
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/topic_guides/blocking/blocking_rules.md
--------------------------------------------------------------------------------
frac
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/topic_guides/performance/optimising_duckdb.md
--------------------------------------------------------------------------------
frac
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/topic_guides/comparisons/phonetic.md
--------------------------------------------------------------------------------
AE
ComparisonLevels
GN
Metaphone
NER
PN
Soundex
WH
WR
XMT
damerau
jaro
levenshtein
winkler
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/topic_guides/comparisons/regular_expressions.md
--------------------------------------------------------------------------------
ComparisonLevels
PQ
RL
UZ
jaro
levenshtein
standardized
substring
substrings
winkler
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/topic_guides/comparisons/comparators.md
--------------------------------------------------------------------------------
frac
jaro
levenshtein
lfloor
rfloor
textsf
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/comparison_helpers.md
--------------------------------------------------------------------------------
splink
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/settings_dict_guide.md
--------------------------------------------------------------------------------
bayes
jaro
probabalistic
splink
sql
sqlglot
tf
winkler
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/linker.md
--------------------------------------------------------------------------------
num
roc
sql
tf
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/comparison_level_composition.md
--------------------------------------------------------------------------------
splink
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/linkereval.md
--------------------------------------------------------------------------------
roc
tf
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/index.md
--------------------------------------------------------------------------------
ADR
BibTeX
DASD
Doidge
Gateshead
Harron
Hepworth
Lewisham
Manassis
MoJ
Slade
Sunter's
TQyNLJRjnhsfJy
UQ
Zd
customizations
doi
lowercased
scalable
sdt
standardized
url
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/CONTRIBUTING.md
--------------------------------------------------------------------------------
PRs
dev
docstrings
repos
splink
--------------------------------------------------------------------------------

Misspelled words:
<context> docs/linkerbloc.md
--------------------------------------------------------------------------------
num
--------------------------------------------------------------------------------

!!!Spelling check failed!!!
```

</details>

> the spellchecker runs for you (spellchecks the docs) but doesn't pass (finds spelling errors)?

Correct, and a lot of them, so I figured that wasn't working correctly. And, I'm pretty sure it's not using the `.dic` and `.aff` files on my computer -- how sure are you that it is using them on yours?

> Would you also be able to tell us which Mac-specific parts you commented out?

I commented out these lines, since they use Homebrew, which is Mac-specific:

https://github.com/moj-analytical-services/splink/blob/1d942f0614817fce7c0f50728d451e9bc86fe83d/scripts/pyspelling/spellchecker.sh#L13-L15

I replaced them with conda package installation (aspell and go-yq), which is cross-platform. Happy to contribute this if you are interested. Would you be open to _switching_ to that, or would you want to have both?

> This is being done as part of this PR - I will update the description to reflect that.

Ah okay, I think this was the crux of my misunderstanding. I was working on doc updates in #2083 and thought I needed to get this spellcheck passing. The name of this PR indicated to me that the feature was _done_, just not _documented_.

zslade commented:
> I get `282`. The full output:

Thank you. Seems like the updates to the custom dictionary are working.

> Correct, and a lot of them, so I figured that wasn't working correctly. And, I'm pretty sure it's not using the `.dic` and `.aff` files on my computer -- how sure are you that it is using them on yours?

So it should be using the .dic file as a dictionary of words to check against. This started off as the one provided by LibreOffice but gets updated with words from the custom dictionary.

If you add a word to `custom_dictionary.txt` and run the spellchecker, you should hopefully see `dictionary.dic` get modified. This is what is happening for m so I’m confident it’s working on my machine at least. Let me know if this doesn’t happen for you.

(I noticed the .dic file wasn’t being saved in the pyspelling directory thought so have now moved it)

> I commented out these lines, since they use Homebrew, which is Mac-specific:
> 
> https://github.com/moj-analytical-services/splink/blob/1d942f0614817fce7c0f50728d451e9bc86fe83d/scripts/pyspelling/spellchecker.sh#L13-L15
> 
> I replaced them with conda package installation (aspell and go-yq), which is cross-platform. Happy to contribute this if you are interested. Would you be open to _switching_ to that, or would you want to have both?

I’ve discussed this with @ThomasHepworth and we think the best way forward would be to put system-specific guidance in the documentation for the installs of aspell etc and remove these steps from the bash script. The bash script will then just run the spellchecker. (We don’t as use conda so a conda only option wouldn’t work for us). I’ll make an issue for this so feel free to pick it up and make a separate PR if you’d like to contribute :)

> Ah okay, I think this was the crux of my misunderstanding. I was working on doc updates in #2083 and thought I needed to get this spellcheck passing. The name of this PR indicated to me that the feature was _done_, just not _documented_.

Sorry for the confusion. The spellchecker functionality had already been added but hadn't actually been _run_ across all docs, and to be honest, hasn’t _yet_ become an established of our Docs workflow. This is what this PR is hoping to address by adding guidance to the docs. But I agree the name of the PR is a bit misleading so I’ll update that!

Thanks again for your input!

ThomasHepworth commented:
One of the output blocks in [regular_expressions.md](https://moj-analytical-services.github.io/splink/topic_guides/comparisons/regular_expressions.html?h=regular#example-using-valid_string_pattern) appears to be incorrect.

![Screenshot 2024-03-28 at 14 15 17](https://github.com/moj-analytical-services/splink/assets/45356472/799162e6-ee39-45f8-96ad-66aa22bd0c27)

ThomasHepworth commented on docs/dev_guides/changing_splink/contributing_to_docs.md:
```diff
@@ -0,0 +1,58 @@
+
+otherwise, PySpelling will printout the spelling mistakes found in each file.
+
+Correct spellings of words not found in a standard dictionary (e.g. Splink) can be recorded as such by adding them to `scripts/pyspelling/custom_dictionary.txt`. (Don't worry about adding them in alphabetical order or accidental duplication as this will be handled automatically by a GitHub Action future.)
```
Link to this?

zslade replied:
Will change is next PR

ThomasHepworth commented on docs/dev_guides/changing_splink/contributing_to_docs.md:
````diff
@@ -0,0 +1,58 @@
+
+To run the spellchecker on either a single markdown file or folder of markdown files, you can use the following script:
+
+```sh
+source scripts/pyspelling/spellchecker.sh <path_to_file_or_folder>
+```
````
Should we also note that this only works on mac at present and further instructions for other operating systems will be released later?

zslade replied:
Good point, will update

ThomasHepworth reviewed (changes requested):
Two minor comments and then you can merge this.

ThomasHepworth reviewed (approved):
This is really great, thank you so much for pulling it together! Sorry it was so much work.
